tests/enrol_advanced_testcase_test.php

```diff
@@ -0,0 +1,256 @@
+      $evento_plugin = enrol_get_plugin($plugin);
+      $courses = $DB->get_recordset_select('course', 'category > 0', null, '', 'id');
+      foreach ($courses as $course)
+      {
+        $instanceid = null;
+        $instances = enrol_get_instances($course->id, true);
+        foreach ($instances as $inst)
+        {
+          if ($inst->enrol == $plugin)
+          {
+            $instanceid = (int)$inst->id;
+            break;
+          }
+        }
+        if (empty($instanceid))
+        {
+          $instanceid = $evento_plugin->add_default_instance($course);
+          if (empty($instanceid))
+          {
+            $instanceid = $evento_plugin->add_instance($course);
+          }
+        }
+        if (!empty($instanceid))
+        {
+          // Do additional config of instance if needed.
+          ($instanceid);
+        }
+      }
+
+      $record = new stdClass();
+      $record->id = 246003;
+      $record->customint1 = 0;
+      $DB->update_record('enrol', $record, $bulk=false);
+      /*Enrol Users into courses*/
+      $enrol->user_sync($trace, $courseid =null);
+      /*Get user enrolment record to count enrolments*/
+      $this->get_enroled_user($this->course1->id);
+      $this->assertEquals($this->enrolments, 33, "Einschreibungen");
+    }
+    public function test_update_student_enrolment()
+    {
+      global $DB;
+      $this->resetAfterTest(false);
+      $eventoenrolstate = 20215;
+      $eventopersonid =  141703;
+      /*Get the user records*/
+      $person = $this->locallib->get_users_by_eventoid_exposed($eventopersonid, $isstudent=null);
+      $user = reset($person);
+      /*Get course settings*/
+      $course = $DB->get_record('course', array('idnumber'=>'mod.mmpAUKATE1.HS18_BS.001'));
+      $this->get_enroled_user($course->id);
+
+      /*Delete a user to re-enrol after*/
+      $DB->delete_records('user_enrolments', array('userid'=> current($person)->id));
+      $DB->get_record('user_enrolments', array('userid'=> current($person)->id));
+      $this->assertEquals($DB->count_records('user_enrolments', array('enrolid'=> $this->user_enrolment->id)), 32);
+      /*Get enrolment instance of course*/
+      $instance = $DB->get_record('enrol', array('id' => $this->user_enrolment->id));
+      /*re-enrol user*/
+      $this->locallib->update_student_enrolment_exposed($eventopersonid, $eventoenrolstate, $instance);
+      /*Count users*/
+      $this->assertEquals($DB->count_records('user_enrolments', array('enrolid'=> $this->user_enrolment->id)), 33);
+    }
+    public function test_enrol_teacher()
+    {
+      sleep(20);
+      global $DB;
+      $eventopersonid =  117828;
+      $this->resetAfterTest(false);
+      /*Get the teacher record*/
+      $person_teacher = $this->locallib->get_users_by_eventoid_exposed($eventopersonid, $isstudent=null);
+      /*Get the course settings*/
+      $course = $DB->get_record('course', array('idnumber'=>'mod.mmpAUKATE1.HS18_BS.001'));
```

Comment:
Wir gehen davon aus, dass in Kurs mit der idnumber mod.mmpAUKATE1.HS18_BS.001 existiert.
Wieso haben wir so einen Kurs? 
Wie garantieren wir, dass wir bei jedem Testdurchlauf so einen Kurs haben?

Reply:
Ich sehe gerade, dass dieser Kurs durch die Funktion `create_moodle_course` erzeugt wird.
Aber wenn ich nur diese Testfunktion aufrufe (und nicht vorher test_user_sync), ist dann dieser Kurs auch in der Datenbank?

Es sollte nicht sein, dass die Testausführung durch die Reihenfolge der Aufrufe beeinträchtigt ist. oder anders gesagt: Jede Testmethode sollte sich für sich alleine aufrufen lassen.